### PR TITLE
Update section 2.4 of 0x11-V2-Authentication.md

### PR DESCRIPTION
**Pull request by jmanico**
adding parity with owasp cheatsheet

This Pull Request relates to issue #...

### CODE REVIEW

**Comment by Sjord** [comment]
> Verify that one of the following password hashing functions are used when storing passwords: argon2id, scrypt, bcrypt or PBKDF2.

"one" is singular, and "are" is plural. So this should be "one of the following password hashing functions is used"

File: 4.0/en/0x11-V2-Authentication.md [hunk 1]
```diff
@@ -79,17 +79,19 @@ Note: Passwords are not to have a maximum lifetime or be subject to password rot
...
 | **2.4.2** | [DELETED] | | | | | |
-| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, typically at least 310,000 iterations. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
-| **2.4.4** | [MODIFIED] Verify that if bcrypt is used, the work factor SHOULD be as large as verification server performance will allow, typically at least 10. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
+| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, with a minimum of at least 310,000 iterations using HMAC-SHA-256 as the internal hash function. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
+| **2.4.4** | [MODIFIED] Verify that if bcrypt is used, the work factor SHOULD be as large as verification server performance will allow, with a minimum of at least 10. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
```
Comment (Sjord) [inline 4.0/en/0x11-V2-Authentication.md:91]: I think "with a minimum of" and "at least" is saying the same thing twice.

File: 4.0/en/0x11-V2-Authentication.md [hunk 2]
```diff
@@ -79,17 +79,19 @@ Note: Passwords are not to have a maximum lifetime or be subject to password rot
...
+| **2.4.1** | [MODIFIED] Verify that one of the following password hashing functions are used when storing passwords: argon2id, scrypt, bcrypt or PBKDF2. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
 | **2.4.2** | [DELETED] | | | | | |
-| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, typically at least 310,000 iterations. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
-| **2.4.4** | [MODIFIED] Verify that if bcrypt is used, the work factor SHOULD be as large as verification server performance will allow, typically at least 10. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
+| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, with a minimum of at least 310,000 iterations using HMAC-SHA-256 as the internal hash function. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
```
Comment (Sjord) [inline 4.0/en/0x11-V2-Authentication.md:90]: I am not sure we should require HMAC-SHA-256. I think HMAC-SHA-1 is also acceptable.

**Comment by jmanico** [comment]
> > Verify that one of the following password hashing functions are used when storing passwords: argon2id, scrypt, bcrypt or PBKDF2.
> 
> "one" is singular, and "are" is plural. So this should be "one of the following password hashing functions is used"

I updated the PR, thank you.

File: 4.0/en/0x11-V2-Authentication.md [hunk 3]
```diff
@@ -79,17 +79,19 @@ Note: Passwords are not to have a maximum lifetime or be subject to password rot
...
+| **2.4.4** | [MODIFIED] Verify that if bcrypt is used, the work factor SHOULD be as large as verification server performance will allow, with a minimum of at least 10. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
 | **2.4.5** | [DELETED] | | | | | |
+| **2.4.6** | [MODIFIED] Verify that if argon2id is used, the configuration SHOULD be as strong as verification server performance will allow, with a minimum configuration of 15 MiB of memory, an iteration count of 2, and 1 degree of parallelism. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
```
Comment (Sjord) [inline 4.0/en/0x11-V2-Authentication.md:93]: I am not a fan of "SHOULD". It sounds like you should do it, but it's not required. But all these requirements are required.

Comment (Sjord) [inline 4.0/en/0x11-V2-Authentication.md:93]: Shouldn't this be `[ADDED]` instead `[MODIFIED]`?

File: 4.0/en/0x11-V2-Authentication.md (hunk 1 above)
Comment (jmanico) [reply 4.0/en/0x11-V2-Authentication.md]: I updated the PR, thank you!

File: 4.0/en/0x11-V2-Authentication.md (hunk 3 above)
Comment (jmanico) [reply 4.0/en/0x11-V2-Authentication.md:93]: I would even say MUST, If you are going to use Argon2id THEN you NEED to use at least these settings....

Comment (jmanico) [reply 4.0/en/0x11-V2-Authentication.md]: Yes and its not fixed!

File: 4.0/en/0x11-V2-Authentication.md (hunk 2 above)
Comment (jmanico) [reply 4.0/en/0x11-V2-Authentication.md]: PS: Send me your email via jim@manicode.com and I'll add you to the ASVS lead calendar invite sessions.

Comment (jmanico) [reply 4.0/en/0x11-V2-Authentication.md]: That's a good point. The Hashcat researchers told me:

PBKDF2-HMAC-SHA1: 720,000 iterations
PBKDF2-HMAC-SHA256: 310,000 iterations
PBKDF2-HMAC-SHA512: 120,000 iterations

Can I just list one of these for ASVS? Or any other wordsmithing suggestions? list all 3?

Comment (jmanico) [reply 4.0/en/0x11-V2-Authentication.md]: I updated this to:

| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, with a minimum of 720,000 iterations with PBKDF2-HMAC-SHA1, a minimum of 310,000 iterations using PBKDF2-HMAC-SHA-256, or with a minimum of 120,000 iterations with PBKDF2-HMAC-SHA-512. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |**

Better?

File: 4.0/en/0x11-V2-Authentication.md [hunk 4]
```diff
@@ -87,7 +87,7 @@ This section cannot be penetration tested, so controls are not marked as L1. How
 | :---: | :--- | :---: | :---:| :---: | :---: | :---: |
 | **2.4.1** | [MODIFIED] Verify that one of the following password hashing functions is used when storing passwords: argon2id, scrypt, bcrypt or PBKDF2. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
 | **2.4.2** | [DELETED] | | | | | |
-| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, with a minimum of 310,000 iterations using HMAC-SHA-256 as the internal hash function. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
+| **2.4.3** | [MODIFIED] Verify that if PBKDF2 is used, the iteration count SHOULD be as large as verification server performance will allow, with a minimum of 720,000 iterations with PBKDF2-HMAC-SHA1, a minimum of 310,000 iterations using PBKDF2-HMAC-SHA-256, or with a minimum of 120,000 iterations with PBKDF2-HMAC-SHA-512. ([C6](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 916 | 5.1.1.2 |
```
Comment (cmlh) [inline 4.0/en/0x11-V2-Authentication.md:90]: What is the outcome if the legacy hardware is unable to support the minimum number of iterations?

I'd prefer that PBKDF2-HMAC was listed as separate requirements for SHA1, SHA256 and SHA512.

Please remove the "-" character between SHA and 1, 256 and 512 too?

Comment (elarlang) [reply 4.0/en/0x11-V2-Authentication.md:90]: https://github.com/OWASP/ASVS/pull/1492

**Pull request merged**